multiview/cotraining/ctclassifier.py

```diff
@@ -0,0 +1,364 @@
+    data which satisfy the 3 conditions for multi-view co-training
+    (sufficiency, compatibility, conditional independence) as detailed in [1].
+
+    Parameters
+    ----------
+    estimator1 : classifier object
+        The classifier object which will be trained on view 1 of the data.
+        This classifier should support the predict_proba() function so that
+        classification probabilities can be computed and co-training can be
+        performed effectively.
+
+    estimator2 : classifier object
+        The classifier object which will be trained on view 2 of the data.
+        Does not need to be of the same type as estimator1, but should support
+        predict_proba().
+
+    Attributes
+    ----------
+    estimator1 : classifier object
+        The classifier used on view 1.
+
+    estimator2 : classifier object
+        The classifier used on view 2.
+
+    class_name: string
+        The name of the class: CTClassifier.
+
+    n_views_ : int
+        The number of views supported by the multi-view classifier
+
+    classes_ : array-like of shape (n_classes,)
+        Unique class labels.
+
+    n_classes_ : int
+        Number of unique classes.
+
+    p_ : int
+        Number of positive examples (second label in classes_) to pull from
+        unlabeled_pool and give "label" at each training round. When the
+        labels are 0 or 1, positive is defined as 1, and in general, positive
+        is the larger label.
+
+    n_ : int
+        Number of negative examples (second label in classes_) to pull from
+        unlabeled_pool and give "label" at each training round. when the
+        labels are 0 or 1, negative is defined as 0, and in general, negative
+        is the smaller label.
+
+    unlabeled_pool_size_ : int
+        Size of pool of unlabeled_pool examples to classify at each iteration.
+
+    num_iter_ : int
+        Maximum number of training iterations to run.
+
+    References
+    ----------
+    [1] Blum, A., & Mitchell, T. (1998, July). Combining labeled and
+    unlabeled_pool data with co-training. In Proceedings of the eleventh
+    annual conference on Computational learning theory (pp. 92-100). ACM.
+
+    """
+
+    def __init__(
+                 self,
+                 estimator1=None,
+                 estimator2=None,
+                 random_state=0
+                 ):
+
+        super().__init__(estimator1, estimator2)
+
+        # if not given, set classifiers as gaussian naive bayes estimators
+        if self.estimator1 is None:
+            self.estimator1 = GaussianNB()
+        if self.estimator2 is None:
+            self.estimator2 = GaussianNB()
+
+        # verify that estimator1 and estimator2 have predict_proba
+        if (not hasattr(self.estimator1, 'predict_proba')
+                or not hasattr(self.estimator2, 'predict_proba')):
+            raise AttributeError("Co-training classifier must be initialized "
```

Comment:
Was there a reason why they must be probabilistic? Like why not treat 1 and 0 as probabilities?

Reply:
To keep things simple let's just leave it as is. Maybe nice to include a todo comment if what I mentioned above is feasible.

Reply:
Based on the Blum and Mitchel Co-training paper, the estimators should be able to assign probabilities to the predictions so that at each round it can choose the "most confident" N to add to the labeled pool. There might be later research describing how to get around this requirement, but for now, since this just implements their paper's algorithm, I'm going to keep it as is.